x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java

```diff
@@ -796,10 +799,26 @@ Collection<Object> createComponents(
         );
         this.persistentTasksService.set(persistentTasksService);
 
+        final boolean autoCreateMainSecurityIndex = DiscoveryNode.isStateless(settings);
         systemIndices.getMainIndexManager().addStateListener((oldState, newState) -> {
-            // Only consider applying migrations if it's the master node and the security index exists
-            if (clusterService.state().nodes().isLocalNodeElectedMaster() && newState.indexExists()) {
-                applyPendingSecurityMigrations(newState);
+            if (clusterService.state().nodes().isLocalNodeElectedMaster()) {
+                // Only consider applying migrations if it's the master node and the security index exists
+                if (newState.indexExists()) {
+                    applyPendingSecurityMigrations(newState);
+                    // In serverless-mode, auto-create the main index eagerly since the security index is always required and should
+                    // be available as quickly as possible
+                } else if (autoCreateMainSecurityIndex && mainSecurityIndexAutoCreationInProgress.compareAndSet(false, true)) {
+                    systemIndices.getMainIndexManager().prepareIndexIfNeededThenExecute(ex -> {
+                        logger.warn("Failed to auto-create security index", ex);
```

Comment:
Should handle "already created" and possibly other "benign" exceptions here to avoid logging as warns.